### DISCUSSION TITLE
xrepo调用conan下载错误

### DISCUSSION THREAD

**Discussion by penguin-madagascar** [xmake-io/xmake#3839] [Q&A] [answered]
我希望在代码中使用特定的conan库，比如zlib：
```
add_rules("mode.debug", "mode.release")

add_requires("conan::zlib")

target("luatest")
    set_kind("binary")
    add_files("src/*.cpp")
    add_packages("conan::zlib")
```
在运行后，会出现如下输出：

> Executing task: xmake -w 
> 
> checking for Microsoft Visual Studio (x64) version ... 2022
> note: install or modify (m) these packages (pass -y to skip confirm)?
> in conan:
>   -> conan::hello latest [vs_runtime:"MT"]
> please input: y (y/n/m)
> 
>   => install conan::hello latest .. failed 
> 
> ERROR: Invalid setting 'MT' is not a valid 'settings.compiler.runtime' value.
> Possible values are ['static', 'dynamic']
> Read "http://docs.conan.io/2/knowledge/faq.html#error-invalid-setting"
> if you want to get more verbose errors, please see:
>   -> D:\luatest\build\.packages\c\conan_hello\latest\cache\installdir.failed\logs\install.txt
> error: install failed!

即MT无法作为compiler.runtime，
这是我的conan设置，可以看到compiler.runtime被设置为dynamic

```
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=14
compiler.runtime=dynamic
compiler.version=193
os=Windows
```

如果在lua文件中强制指定runtime为dynamic，会导致msvc无法使用
有没有办法阻止xmake对runtime的修改？

**Comment by penguin-madagascar** [comment]
错误中的conan::hello是自建package，已确认通过conan可以正常安装
在xrepo中调用任意conan package(甚至不存在的package)，都会出现相同的错误

**Comment by waruqi** [comment]
贴下 conan 版本

**Reply by penguin-madagascar** [reply]
Conan version 2.0.6

**Comment by waruqi** [comment]
还有带 -vD 完整logs

**Reply by penguin-madagascar** [reply]
控制台输入：
`xrepo install -vD conan::zlib`
运行结果：
> xmake f -c --require=n -vD
> checking for platform ... windows
> checking for architecture ... x64
> checking for cl.exe ... C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.36.32532\bin\HostX64\x64\cl.exe
> checking for Microsoft Visual Studio (x64) version ... 2022
> checkinfo: cannot runv(zig.exe version), No such file or directory
> checking for zig ... no
> checkinfo: cannot runv(zig.exe version), No such file or directory
> checking for zig ... no
> checking for link.exe ... C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.36.32532\bin\HostX64\x64\link.exe
> checking for the linker (ld) ... link.exe
> configure
> {
>     kind = static
>     plat = windows
>     arch = x64
>     buildir = build
>     ccache = true
>     host = windows
>     ndk_stdcxx = true
>     clean = true
>     mode = release
>     vs = 2022
> }
> xmake require -v -D -j 6 --extra={system=false} conan::zlib
> checkinfo: cannot runv(unzip.exe -v), No such file or directory
> checking for unzip ... no
> checking for 7z ... C:\Program Files\xmake\winenv\bin\7z
> checking for git ... ok
> checkinfo: cannot runv(gzip.exe --version), No such file or directory
> checking for gzip ... no
> finding zlib from conan ..
> checking for conan::zlib ... no
> note: install or modify (m) these packages (pass -y to skip confirm)?
> in conan:
>   -> conan::zlib latest [vs_runtime:"MT"]
> please input: y (y/n/m)
> 
> installing zlib from conan ..
> checking for conan ... ok
> conan config install "C:\\Program Files\\xmake\\scripts\\conan"
> Copying file xmake_generator.py to C:\Users\jiangxingda\.conan2\extensions\generators
> generate C:\Users\JIANGX~1\AppData\Local\Temp\.xmake\230613\xrepo\working\build\.conan\zlib\conanfile.txt ..
> conan install . -g XmakeGenerator --profile:build=profile_build.txt --profile:host=profile_host.txt --build=missing
> ERROR: Invalid setting 'MT' is not a valid 'settings.compiler.runtime' value.
> Possible values are ['static', 'dynamic']
> Read "http://docs.conan.io/2/knowledge/faq.html#error-invalid-setting"
> error: @programdir\core\sandbox\modules\os.lua:378: execv(conan install . -g XmakeGenerator --profile:build=profile_build.txt --profile:host=profile_host.txt --build=missing) failed(1)
> stack traceback:
>     [C]: in function 'error'
>     [@programdir\core\base\os.lua:913]:
>     [@programdir\core\sandbox\modules\os.lua:378]:
>     [@programdir\core\sandbox\modules\os.lua:291]: in function 'vrunv'
>     [...dir\modules\package\manager\conan\v2\install_package.lua:324]:
>     [...ramdir\modules\package\manager\conan\install_package.lua:41]:
>     [@programdir\modules\package\manager\install_package.lua:110]:
>     [@programdir\core\package\package.lua:2490]: in function 'script'
>     [...dir\modules\private\action\require\impl\utils\filter.lua:125]: in function 'call'
>     [...\modules\private\action\require\impl\actions\install.lua:337]:
> 
>   => install conan::zlib latest .. failed
> error: @programdir\modules\private\async\runjobs.lua:256: ...\modules\private\action\require\impl\actions\install.lua:469: install failed!
> stack traceback:
>     [C]: in function 'error'
>     [@programdir\core\base\os.lua:913]:
>     [...\modules\private\action\require\impl\actions\install.lua:469]: in function 'catch'
>     [@programdir\core\sandbox\modules\try.lua:123]: in function 'try'
>     [...\modules\private\action\require\impl\actions\install.lua:328]:
>     [...modules\private\action\require\impl\install_packages.lua:479]: in function 'jobfunc'
>     [@programdir\modules\private\async\runjobs.lua:232]:
> 
> stack traceback:
>         [C]: in function 'error'
>         @programdir\core\base\os.lua:913: in function 'base/os.raiselevel'
>         (...tail calls...)
>         @programdir\modules\private\async\runjobs.lua:256: in field 'catch'
>         @programdir\core\sandbox\modules\try.lua:123: in global 'try'
>         @programdir\modules\private\async\runjobs.lua:218: in upvalue 'cotask'
>         @programdir\core\base\scheduler.lua:404: in function <@programdir\core\base\scheduler.lua:397>
> error: @programdir\core\main.lua:300: @programdir\core\sandbox\modules\os.lua:378: execv(xmake require -v -D -j 6 --extra={system=false} conan::zlib) failed(-1)
> stack traceback:
>     [C]: in function 'error'
>     [@programdir\core\base\os.lua:913]:
>     [@programdir\core\sandbox\modules\os.lua:378]:
>     [@programdir\modules\private\xrepo\action\install.lua:300]: in function '_install_packages'
>     [@programdir\modules\private\xrepo\action\install.lua:307]:
>     [@programdir\modules\private\xrepo\main.lua:196]:
>     [@programdir\plugins\lua\main.lua:123]:
>     [C]: in function 'xpcall'
>     [@programdir\core\base\utils.lua:280]:
>     [@programdir\core\base\task.lua:501]: in function 'run'
>     [@programdir\core\main.lua:298]: in function 'cotask'
>     [@programdir\core\base\scheduler.lua:404]:
> 
> stack traceback:
>         [C]: in function 'error'
>         @programdir\core\base\os.lua:913: in function 'base/os.raiselevel'
>         (...tail calls...)
>         @programdir\core\main.lua:300: in upvalue 'cotask'
>         @programdir\core\base\scheduler.lua:404: in function <@programdir\core\base\scheduler.lua:397>

**Comment by waruqi** [comment, marked as answer]
这是因为 conan 2.0 改动了很多东西，最近刚重新适配，部分设置可能没兼容上。

我改进了下，再试试，xmake update -s dev

**Reply by penguin-madagascar** [reply]
感谢，可以正常使用了
